### PR TITLE
SMTH-7: Simulation end logic

### PR DESCRIPTION
**Pull request by dbarton-aoe**
Now, when one of the teams is knocked out is simulation state changes to complete to prevent it from running forever. This was done by listening to each actor's KnockedOut event, at which time the active teams variable was updated in the EncounterCtrl. If all members of a team were KO'd, the sim ends.

Also updated the actor behaviourCtrl APIs to simplify and streamline the processing of configuration data. Behaviours now pass configs through the decorator pattern to adjust values and make changes that the ActorCtrl can then process.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Encounters/EncounterCtrl.cs [hunk 1]
```diff
@@ -64,6 +70,68 @@ void Update()
...
+        }
+    }
+
+    public void AddActor(ActorCtrl actor)
+    {
+        if (m_actors != null && !m_actors.Contains(actor))
+        {
+            m_actors.Add(actor);
+        }
+    }
```
Comment (dbarton-aoe) [inline UnityProject/Assets/Scripts/Encounters/EncounterCtrl.cs:134]: These functions were just moved, so these changes can be ignored.

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs [hunk 2]
```diff
@@ -26,12 +27,56 @@ public class ActorCtrl : MonoBehaviour
...
+        get
+        {
+            bool isActive = false;
+            if (m_actorData != null)
+            {
+                isActive = m_actorData.Health > 0;
+            }
+            return isActive;
+        }
+    }
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:53]: Just curious, in C#, why use a getter for a piece of code big enough to be a method?

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs [hunk 3]
```diff
@@ -40,6 +85,14 @@ private void Awake()
         m_behaviourCtrl.m_debugKeyCode = m_debugKeyCode;
     }
 
+    private void Start()
+    {
+        if (ActorData != null)
+        {
+            ActorData.HealthAdjusted += OnHealthChanged;
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:92]: in c# 8, you could od ActorData?.Hea..... and remove the if :)

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs [hunk 4]
```diff
@@ -58,11 +111,28 @@ public void EnqueueAction(ActorActionCtrl action)
 
     public void UpdateActor(EncounterCtrl encounterCtrl)
     {
+        SpeedConfig speedConfig = null;
+
+        // allow our behaviour to update
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:116]: Comment can probably be removed. The if code is pretty clear.

File: UnityProject/Assets/Scripts/Encounters/EncounterCtrl.cs [hunk 5]
```diff
@@ -64,6 +70,68 @@ void Update()
             m_manualStepTriggered = Input.GetKeyDown(m_stepKeyCode);
         }
     }
+   
+    void StartEncounter()
+    {
+        m_encounterState = EncounterState.Running;
+        m_runningCRT = StartCoroutine(RunEncounter());
```
Comment (richardlalancette) [inline UnityProject/Assets/Scripts/Encounters/EncounterCtrl.cs:77]: neat :)

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs (hunk 2 above)
Comment (dbarton-aoe) [reply UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:53]: Properties are the defacto getters and setters in C#, so just following the convention here.

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs (hunk 3 above)
Comment (dbarton-aoe) [reply UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:92]: Hm, tried it, but got this error: `The left-hand side of an assignment must be a variable, property or indexer`. I guess this doesn't work when assigning a delegate function to an event.

File: UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs (hunk 4 above)
Comment (dbarton-aoe) [reply UnityProject/Assets/Scripts/Encounters/ActorCtrl.cs:116]: Will do. Have another PR coming with this change 👍

**Pull request merged**